portal/src/graphql/portal/SingleSignOnConfigurationWidget.tsx

```diff
@@ -0,0 +1,375 @@
+import React, {
+  Dispatch,
+  SetStateAction,
+  useCallback,
+  useContext,
+  useEffect,
+  useState,
+} from "react";
+import { Toggle, Label, TextField } from "@fluentui/react";
+import { FormattedMessage, Context } from "@oursky/react-messageformat";
+
+import ExtendableWidget from "../../ExtendableWidget";
+import {
+  OAuthSSOProviderConfigState,
+  SingleSignOnScreenState,
+} from "./SingleSignOnConfigurationScreen";
+import { OAuthSSOProviderType } from "../../types";
+import {
+  errorFormatter,
+  makeMissingFieldSelector,
+  Violation,
+  violationSelector,
+} from "../../util/validation";
+
+import styles from "./SingleSignOnConfigurationWidget.module.scss";
+
+interface WidgetHeaderLabelProps {
+  icon: React.ReactNode;
+  serviceMessageId: string;
+}
+
+interface WidgetHeaderProps extends WidgetHeaderLabelProps {
+  enabled: boolean;
+  setEnabled: (event: any, enabled?: boolean) => void;
+}
+
+interface SingleSignOnConfigurationWidgetProps {
+  className?: string;
+  serviceProviderType: OAuthSSOProviderType;
+  screenState: SingleSignOnScreenState;
+  setScreenState: Dispatch<SetStateAction<SingleSignOnScreenState>>;
+  violations?: Violation[];
+}
+
+type WidgetTextFieldKey = keyof Omit<
+  Omit<OAuthSSOProviderConfigState, "enabled">,
+  "type"
+>;
+
+type WidgetErrorState = Partial<Record<WidgetTextFieldKey, string>>;
+
+const TEXT_FIELD_STYLE = { errorMessage: { whiteSpace: "pre" } };
+
+const serviceProviderMessageId: { [key in OAuthSSOProviderType]: string } = {
```

Comment:
I would recommend combining the maps into a single map with a descriptor interface as value, e.g.:
```typescript
interface OAuthProviderDef {
  id: string;
  iconNode: React.ReactNode;
  fields: Set<string>;
}
const oauthProviders: Record<OAuthSSOProviderType, OAuthProviderDef>;
```